src/util/fileinfo.h

```diff
@@ -0,0 +1,304 @@
+#pragma once
+
+#include <gtest/gtest_prod.h>
+
+#include <QDateTime>
+#include <QDir>
+#include <QFile>
+#include <QFileInfo>
+#include <QUrl>
+#include <QtDebug>
+
+#include "util/assert.h"
+
+namespace mixxx {
+
+/// Helper functions and extensions for QFileInfo using Mixxx terminology.
+namespace fileinfo {
+
+/// Only intended for internal usage.
+inline const QFileInfo& assertNotRelative(const QFileInfo& fileInfo) {
+    // Detect potential, unintended side-effects by the outer context,
+    // i.e. when accessing the transient working directory to resolve
+    // of relative paths. The path should either be empty or absolute,
+    // i.e. not relative.
+    DEBUG_ASSERT(!fileInfo.isRelative());
+    return fileInfo;
+}
+
+/// Returns the permanent location of a file.
+inline QString location(const QFileInfo& fileInfo) {
+    return assertNotRelative(fileInfo).absoluteFilePath();
+}
+
+/// The directory part of the location, i.e. excluding the file name.
+inline QString locationPath(const QFileInfo& fileInfo) {
+    return assertNotRelative(fileInfo).absolutePath();
+}
+
+/// Returns the current location of a physical file, i.e.
+/// without aliasing by symbolic links and without any redundant
+/// relative paths.
+inline QString canonicalLocation(const QFileInfo& fileInfo) {
+    return assertNotRelative(fileInfo).canonicalFilePath();
+}
+
+/// The directory part of the canonical location, i.e. excluding the file name.
+inline QString canonicalLocationPath(const QFileInfo& fileInfo) {
+    return assertNotRelative(fileInfo).canonicalPath();
+}
+
+/// Decide if two canonical locations have a parent/child
+/// relationship, i.e. if the sub location is contained
+/// within the tree originating from the root location.
+///
+/// Both canonical locations must not be empty, otherwise
+/// false is returned.
+bool isRootSubCanonicalLocation(
+        const QString& rootCanonicalLocation,
+        const QString& subCanonicalLocation);
+
+} // namespace fileinfo
+
+/// A thin wrapper (shim) around QFileInfo with additional methods.
```

Comment:
Why not inherit from QFile info if we just  want to add functions?
This way we get rid of a log of boilerplate code.

Reply:
We don't want to expose all functions unconditionally as I explained. And we cannot override any existing, non-virtual functions. QFileInfo is not made for being inherited from, the even the destructor is non-virtual.

Reply:
OK. I have no strong opinion here.
Please add add the function you don't want as a comment.
Or at least explain the issue somehow.
Can't we just delete them in an inherited class?
That would document it well.

Reply:
Using inheritance here is wrong. Both conceptually (Liskov) and technically.